add client support for explicit tls

bellini666 opened the pull request:
Hi @pohmelie ,

I saw the discussion on https://github.com/aio-libs/aioftp/issues/37 and decided to create this PR based on @sammichaels fork.

The only thing I did here was to merge the fork on top of the upstream master branch and adjust the conflicts.

Please let me know what I need to do to get this merged and released and I'll adjust it myself, as I currently need explicit TLS support and the code here seems to make it work.

pohmelie commented:
Is it a client-only explicit tls support? I see no new commands in server.py

pohmelie commented on src/aioftp/client.py:
```diff
@@ -137,7 +140,13 @@ def __init__(
+            open_connection,
+            ssl=ssl if not explicit_tls else None,
+            **siosocks_asyncio_kwargs,
+        )
```
I think `ssl` and `explicit_tls` can't be passed together, so we need to check this and raise an error in this case

pohmelie commented on src/aioftp/client.py:
```diff
@@ -636,6 +645,57 @@ async def connect(self, host, port=DEFAULT_PORT):
+        try:
+            await self.command("AUTH TLS", "234")
+        except errors.StatusCodeError:
+            raise errors.TLSError("Explicit TLS is not supported on this server") from None
```
No need to wrap status code error in another. More to say, we are loosing the actual response from server.

pohmelie commented on src/aioftp/client.py:
```diff
@@ -636,6 +645,57 @@ async def connect(self, host, port=DEFAULT_PORT):
+                raise errors.TLSError(
+                    "Connection was upgraded to TLS, but the commands to upgrade the data channel"
+                    f"have failed {e.received_codes=}: {e.info}",
+                )
```
Same here

pohmelie commented on src/aioftp/client.py:
```diff
@@ -658,6 +718,9 @@ async def login(
         :raises aioftp.StatusCodeError: if unknown code received
         """
+        if self.explicit_tls:
+            await self.upgrade_to_tls()
```
Not sure we need this done like. Why it is «explicit» if it is actually «implicit» right now? Lets leave this on user shoulders to call `upgrade_to_tls` if he wants.

pohmelie commented on src/aioftp/client.py:
```diff
@@ -121,6 +123,7 @@ def __init__(
         parse_list_line_custom=None,
         parse_list_line_custom_first=True,
         passive_commands=("epsv", "pasv"),
+        explicit_tls=False,
```
We do not need a parameter for this. Since this is a «explicit» tls command. So the user decide when to call `upgrade_to_tls`

pohmelie commented on src/aioftp/client.py:
```diff
@@ -673,6 +736,15 @@ async def login(
+                raise errors.TLSError(
+                    "Connection was upgraded to TLS, but the commands to upgrade the data channel"
+                    f"have failed {e.received_codes=}: {e.info}",
+                )
```
This part is also should be removed, I think

pohmelie commented on src/aioftp/client.py:
```diff
@@ -983,7 +1055,14 @@ def append_stream(self, destination, *, offset=0):
+        *,
+        write_into=False,
+        block_size=DEFAULT_BLOCK_SIZE,
+    ):
```
No need to reformat things, if linters are ok. The project moved from 88 to 120 line length.

pohmelie commented on src/aioftp/client.py:
```diff
@@ -1050,7 +1129,14 @@ def download_stream(self, source, *, offset=0):
+        *,
+        write_into=False,
+        block_size=DEFAULT_BLOCK_SIZE,
+    ):
```
Same here

pohmelie commented on src/aioftp/client.py:
```diff
@@ -1190,6 +1276,19 @@ async def get_stream(self, *command_args, conn_type="I", offset=0):
             throttles={"_": self.throttle},
             timeout=self.socket_timeout,
         )
+        if self.explicit_tls:
```
Maybe this should be a private attribute with meaning of «upgraded to tls» or something like this

pohmelie commented on src/aioftp/client.py:
```diff
@@ -1190,6 +1276,19 @@ async def get_stream(self, *command_args, conn_type="I", offset=0):
+                    server_hostname=self.server_host,
+                )
+            except ssl.SSLError as e:
+                raise errors.TLSError("Unable to upgrade data connection to TLS") from e
```
No need to wrap error here

pohmelie commented on src/aioftp/common.py:
```diff
@@ -589,3 +601,71 @@ def setlocale(name):
+
+    @property
+    def session(self):
+        return self._session
```
We don't need a property if we expose the mutable object

pohmelie commented on src/aioftp/errors.py:
```diff
@@ -81,3 +82,9 @@ class NoAvailablePort(AIOFTPException, OSError):
     """
+
+
+class TLSError(AIOFTPException):
```
Default errors are more than enough

pohmelie reviewed (changes requested):
Here is what need to be done (if possible):
- [ ] tests: not sure how to test this ssl things actually, since we have no server code)
- [ ] documentation: check if docs are extended with new comments

pohmelie commented:
There is also mechaninc in ftp to remove tls out from connection, so this pr is less than a 1/4 of what need to be done. But, I think we will merge this anyway after review/changes, since I think no one will implement full support till the universe end

bellini666 commented:
Hi @pohmelie 

Thanks for the review

I've done the adjustments you requested, will try to add a couple of tests and docs tomorrow. Let me know if there's any other required adjustment here

pohmelie commented on src/aioftp/client.py:
```diff
@@ -636,6 +640,45 @@ async def connect(self, host, port=DEFAULT_PORT):
+        if self.stream.writer.transport.get_extra_info("ssl_object"):
+            return
+
+        self._upgraded_to_tls = True
```
Lets set this after all commands ends successfully

pohmelie commented on src/aioftp/client.py:
```diff
@@ -636,6 +640,45 @@ async def connect(self, host, port=DEFAULT_PORT):
+        :param sslcontext: custom ssl context
+        :type sslcontext: :py:class:`ssl.SSLContext`
+        """
+        if self.stream.writer.transport.get_extra_info("ssl_object"):
```
We should also inspect if tls upgraded already

pohmelie commented on src/aioftp/client.py:
```diff
@@ -674,6 +717,11 @@ async def login(
+        self.logged_in = True
+
+        if self._upgraded_to_tls:
+            await self._send_tls_protection_commands()
```
I don't think we need to upgrade to tls implicitly at all. What this case is about?

bellini666 replied:
If I understood correctly the author's implementation, this is needed to support the authentication to happen before or after the `upgrade_to_tls`, but those protection commands can only be sent after the login itself.

bellini666 commented:
Hi @pohmelie ,

Friendly ping about this PR :)

Waiting on this to be merged and released to be able to use it in a project at work

pohmelie commented on docs/client_tutorial.rst:
```diff
@@ -375,6 +375,20 @@ to specify global timeout for socket io operations.
+-------------------
+
+Just like Python's `ftplib.FTP_TLS`, aioftp supports TLS upgrade. This is
+done by calling :py:meth:`aioftp.Client.upgrade_tls` after instantiating the
```
Naming missmatch 
```suggestion
done by calling :py:meth:`aioftp.Client.upgrade_to_tls` after instantiating the
```

pohmelie commented on uv.lock:
```diff
@@ -0,0 +1,587 @@
+version = 1
```
This file should be removed

bellini666 replied:
Ohhh sorry, I used `uv` to install and run the tests and accidentally committed this T_T

Removed!

bellini666 commented:
@pohmelie should be all good now! :)

pohmelie commented:
> @pohmelie should be all good now! :)

Checkout the tests for python 3.9

bellini666 commented:
> > @pohmelie should be all good now! :)
> 
> Checkout the tests for python 3.9

Fixed it. Ended up removing the typing info. Although I like type annotations, this project doesn't make use of them, so no need to have them.

pohmelie commented:
Nah, broken again. Did you run tests localy?

bellini666 commented:
> Nah, broken again. Did you run tests localy?

I had before, but not after the typing removal, by bad :(

(the culprit was this: [`146aef5` (#183)](https://github.com/aio-libs/aioftp/pull/183/commits/146aef5e63ccef495282a10009b2cf595e5ba064#diff-0cb34502f4f31cd519adc377a42fa19735a3069781423362a8d21f71d53b3957L652). I removed the `=None` as well by mistake in a hurry)

Fixed now, ran the tests locally to make sure (but not in every python version supported)

pohmelie commented:
Thank you for the contibution @bellini666

bellini666 commented:
@pohmelie thanks mostly to @sammichaels . I just made sure his work on his fork got here :)
